control/injector/Injector.php

```diff
@@ -551,6 +551,14 @@ protected function instantiate($spec, $id=null, $type = null) {
 			$constructorParams = $spec['constructor'];
 		}
 
+		// If we're dealing with a DataObject, pass through Singleton flag as second argument
+		if ($type != 'prototype' && is_subclass_of($class, 'DataObject')) {
+			if (!count($constructorParams)) {
+				$constructorParams[0] = null;
+			}
+			$constructorParams[1] = true;
```

Comment:
Sometimes constructor args are named.

```yaml
MyObject:
  class: MyObject
  constructor:
    record: []
    singleton: true
```

however that's super edge case. Maybe only apply this fix if $constructorParams are totally empty?

Not a review change request, because it feels edge case enough we can ignore it, but worth thinking about before we modify this in 4.x.

Reply:
Could we just convert to numeric keys:
```php
$constructorParams = array_values($constructorParams);
```
The constructor args will be passed by position rather than name right? So I assume it should be safe to discard any non-numeric keys.

Reply:
It's safer to only supply default constructor params, in this case based on `$type != 'prototype' && empty($constructorParams)`. Trying to merge in based on position / key is guesswork at best.

Reply:
I don't think we can add a `empty($constructorParams)` check to this because `$constructorParams = [null]` would not be considered empty, would it?

Reply:
Yes but when would we have `[null]`?

Reply:
Don't know :) I'm just guessing so I'll trust you on this lol. Have added a commit with a `empty($constructorParams)` condition.